### ISSUE TITLE
Migrate to Google Analytics 4 (GA4)

### ISSUE THREAD

**Issue by chalin** [kubernetes/website#35299]
This is part of a CNCF-wide effort to upgrade project websites to GA4 since **Google has deprecated Universal Analytics** (UA). For more details, see:

- https://github.com/cncf/techdocs/issues/108, including reference to:
  - A suggested [migration process](https://github.com/cncf/techdocs/blob/main/docs/analytics.md#migrating-to-google-analytics-4), 
  which we've used to [migrate a few CNCF projects](https://docs.google.com/spreadsheets/d/1Mx4LhdI2Un-rvGMI73SlHxQH9D2HABAJclMB3dd6lnA) already.

### Tasks, stage 1 (no code changes are necessary)

- [x] Create a GA4 site tag using the **GA4 setup assistant**
  from the "**Kubernetes - UA**" (`UA-36037335-10`) analytics console.
  - The new GA4 measurement ID is **`G-JPP6RFM2BP`**. User access is the same as for the original UA site tag.
- [x] Manually add the GA4 site tag to the list of `UA-36037335-10` **[Connected] Site Tags** under **Admin** > **Tracking Info** > [Tracking Code]
- [x] Confirm that the GA4 site tag is receiving events (for a screenshot, see below).
- [x] Confirm that the UA site tag, `UA-36037335-10`, is still receiving events (for a screenshot, see below).

### Tasks, stage 2 (optional, can be done later)

- [x] Manually add `UA-36037335-10` as a connected site-tag of the GA4 tag
- [x] Switch to using GA4 ID:
  - [x] Netlify config - add environment variable (done by @onlydole, thanks!)
  - [x] #36010
  - [x] Wait for this to be deployed to the production server.
  - [x] Confirm that the GA4 site tag is receiving events
  - [x] Confirm that the UA site tag, `UA-36037335-10`, is still receiving events
- [x] Enable GA for existing release-branch deploys
  - [x] Confirm that the GA4 site tag is receiving events
  - [x] Confirm that the UA site tag, `UA-36037335-10`, is still receiving events
- [x] Remove the connection of the GA4 tag from the UA config and ensure events are still being processed:
  - [x] Remove the GA4 tag from the UA config: _Tracking Info > Tracking Code > Connected Site Tags_ section. By doing so, all non-kubernetes.io sites using the UA ID will stop forwarding events to the GA4 ID

### Release-branches and page-feedback

- [x] Release-branches (see issue-comment below for 2022/09 page-view data):
  - [x] 1.21 - #37565
  - [x] 1.22 - #37566
  - [x] 1.23 - #37567
  - [x] 1.24 - #36322
  - [x] ~1.25 - #37669~ The 1.25 release branch had been created by mistake. It will be recreated from main in Dec.
  - [x] 1.25, take 2 - #38494

### Followup tasks

- Enable GA for release-branch deploys
  - #36325
  - #36334
- Remove the connection of the GA4 tag from the UA config and ensure events are still being processed:
  - Remove the GA4 tag from the UA config: _Tracking Info > Tracking Code > Connected Site Tags_ section. By doing so, all non-kubernetes.io sites using the UA ID will stop forwarding events to the GA4 ID
  - Confirm that the GA4 site tag is receiving events
  - Confirm that the UA site tag, `UA-36037335-10`, is still receiving events
- #37657
- #37201
- Backup of UA data: as of 2024Q1 UA data will be deleted: ensure that #sig-docs-maintainers are aware of this so that they can plan, if necessary, a strategy for backing up the data.

### Related

- https://github.com/kubernetes/community/issues/6965
- https://github.com/kubernetes/contributor-site/issues/354
- https://github.com/kubernetes-sigs/release-notes/issues/430
- https://github.com/kubernetes/website/issues/37801
- Generated API reference pages need to have analytics updated too. For questions, see https://github.com/kubernetes/website/issues/35299#issuecomment-1310622168.

/cc @nate-double-u @caniszczyk 

---

Current website analytics info:

- Analytics snippet is added via `website/layouts/partials/head.html` -- for the relevant excerpt, see:
 https://github.com/kubernetes/website/blob/4b435b453bb3c6c06458337ce51ec7b48d7a74b7/layouts/partials/head.html#L9-L19
- Hugo version used is 0.101.0 

---

See also:

- #36011

[release-1.25 branch]: https://github.com/kubernetes/website/tree/release-1.25
[Tracking Code]: https://analytics.google.com/analytics/web/#/a98841695w90600397p94297700/admin/tracking/tracking-code/
[Connected]: https://support.google.com/analytics/answer/9973999

**Comment by a-mccarthy** [comment]
@chalin Thank you for creating this issue and the related resources on the migration. I'm reading through your docs at the moment, but wanted to let you know that I just checked, and I dont have permissions to add users to our current GA account. i've requested access and I would love to help you with this project if you are looking for help :) 

I've worked with @jimangel in the past on getting GA access. Jim, can you help here with getting access? Or helping me get permissions to grant access to folks?

**Comment by chalin** [comment]
Very cool that you've (@sftim) been upgrading Hugo and Docsy ([to v0.2.0 via #32906](https://github.com/kubernetes/website/pull/32906)). On that topic, Docsy is now at [v0.4.0](https://github.com/google/docsy/blob/main/CHANGELOG.md#040), if you consider upgrading, let me know if you have any issues.

**Comment by jimangel** [comment]
Sorry I missed your ping @a-mccarthy! I was chatting with @chalin, @jeefy, and @onlydole on the CNCF slack and provided Google Analytic access to a CNCF email account for completing the work. Leaving this comment for record keeping 😄

**Comment by chalin** [comment]
Progress update: GA4 site tag has been created from the main UA property (`UA-36037335-10`). User access is the same for both. The new GA4 site tag is receiving events and original UA site tag continues to process events.

GA4 screenshot:
> <img width="500" alt="GA4 screenshot" src="https://user-images.githubusercontent.com/4140793/184543019-6f81ae2e-e265-49b4-af81-b448aa4068cb.png">

UA4 screenshot:
> <img width="500" alt="UA4 screenshot" src="https://user-images.githubusercontent.com/4140793/184543065-0b5286a3-9647-4eb5-a107-cc03834be9a9.png">

**Comment by chalin** [comment]
Hi all, please comment on:

- #36011

**Comment by chalin** [comment]
Analytics for release-branch subdomains, July 27 - Aug 25, 2022, FYI:

> <img width="703" alt="image" src="https://user-images.githubusercontent.com/4140793/186846501-83ce1854-92df-4f26-83c0-adff752b12f8.png">

**Comment by a-mccarthy** [comment]
@chalin anything you need help with on this issue? Are you blocked by anything?

**Comment by chalin** [comment]
> @chalin anything you need help with on this issue? Are you blocked by anything?

Thanks @a-mccarthy. I'm just back today, and still catching up. Will reach out with an update as soon as I can.

**Comment by chalin** [comment]
Analytics for release-branch subdomains from Aug 29 - Sept 27:

> <img width="496" alt="image" src="https://user-images.githubusercontent.com/4140793/192888126-e3f95bca-3c58-469d-a7dd-fbfe03bc356f.png">

So to answer @sftim's question from https://github.com/kubernetes/website/pull/36322#issuecomment-1229198223, yes, we'll be updating the GA site tag for release-branches from v1.25 to v1.21. 

Does that sound reasonable?

**Comment by a-mccarthy** [comment]
just wanted to note that we have a google analytics custom event for our feedback yes/no button that shows at the bottom of each page. we should make sure that this works the same in the next version (not sure if we've track this somewhere or not). More details in on the tag in: https://github.com/kubernetes/website/blob/main/layouts/partials/feedback.html

**Comment by chalin** [comment]
@a-mccarthy: how do you access the feedback currently?

**Comment by a-mccarthy** [comment]
@chalin it gets log as an event, so you can view it in google analytics in the Behavior>Events section. I was just looking into this b/c i was curious, and was told of the connection between the feedback on the docs pages and google analytics. 

At the moment, we don't report out this data anywhere, but we should at some point :) 

Heres the code snippet for how it's captured and what we send. 
```
gtag('event', 'click', {
        'event_category': 'Helpful',
        'event_label': window.location.pathname,
        value
      });
```
The value in this case is 1 for yes and 0 for no. And it gets averaged in GA for an average event value.

**Comment by a-mccarthy** [comment]
following up about the feedback form again, it appears that we are seeing some bot data in the analytics (more details in the #37201). It appears that known bot traffic is filtered out by default in GA4, https://support.google.com/analytics/answer/9888366?hl=en. And we can filtering it out in the current GAs. Wanted to post an update here for awareness :)

**Comment by chalin** [comment]
> ... our feedback yes/no button that shows at the bottom of each page. we should make sure that this works the same in the next version

@a-mccarthy: it's not clear that it will work as is. Since page feedback-data processing might have a little less of an urgency than ensuring that GA4 events are being tracked ASAP, I've created a separate issue (and as I mention there, suggest that we tackle it as a followup item to this issue):

- #37657

**Comment by chalin** [comment]
Update:

- All release-branch deploys (as of the time of writing) are using the GA4 tag
- I've removed the GA4 tag from the UA config. See the opening comment for details concerning this step.

**Comment by chalin** [comment]
@sftim @nate-double-u @a-mccarthy et al. : as you might recall, #36010 added a fake GA ID to the Hugo config file with a comment letting builders know that the real GA4 ID is set via a Netlify-build environment variable.

A consequence of this approach is that, for example, when the next release-branch created, someone (the release-lead?) will need to set the GA4 ID environment variable for the new release-branch by copying it from the main site deploy settings. Or, ...

It _might_ be possible to avoid such copying of env. vars because **Netlify has a new way of managing environment variables** that has a concept of **scope and deploy contexts**.

> <img width="924" alt="image" src="https://user-images.githubusercontent.com/4140793/200659312-3c29698c-9ee3-43e8-893a-3de9fce4fc84.png">

Here's the link to the docs from that image: https://docs.netlify.com/environment-variables/overview/ (search, e.g., for scope). I don't have access to the website's Netlify account so I can't explore those settings for Kubernetes. Maybe @onlydole, @a-mccarthy, @jimangel or @sftim can, and let us know what they think is possible?

**Comment by chalin** [comment]
@a-mccarthy @sftim et al.: There are API reference pages like: https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands

I have a few questions about these API reference pages:

- Are we tracking analytics for these pages? They seem to be tracked under UA, but when I look at the page HTML, I don't see any UA ID. Also, UA console says that these pages are actually being hosted by https://jamesdefabia.github.io, but that host reports a 404 for the given paths.
- Do we want these pages to be tracked via GA4?
- Where are these pages generated from?

**Comment by a-mccarthy** [comment]
@chalin and I talked through some open items on this project, specifically around Netlify access and selecting a way to organize properties going forward. I have a summary below or what we talked about and @chalin plans to bring this to the next sig docs meeting as well. (@chalin please chime in if i've miss summarized or missed anything)

Currently @chalin has been manually updating release branches with the new Google Analytics 4 ID. This is not ideal, b/c the process is so manual. This should be something we can configure in Netlify, but someone with access will need to go in and do the testing and configuring. See https://github.com/kubernetes/website/issues/37877 for more details. @chalin has offered to help with this, but needs access. In general, we've had a few requests from folks in the community to access or update things in Netlify in the past few months and we have a Github Discussion around getting the right folks access as well: https://github.com/kubernetes/website/discussions/38019. This will block forward progress on this task without proper access to Netlify. 

Based on @chalin's research the best way and (google) recommended way to set up the new google analytics assets is to create one account and have all websites tied to the account via a single ID. This means that all web entities (kubernetes.io, kubernetes.dev, the minikube etc.) that we want to track analytics under will use the same ID and be collected into the same "pool" of data in Google Analytic. Then, within GA we can create custom reports that display data for the different sites.  The new google analytics has greatly reduced the amount of default reports available to users, even if we created separate organizations and ID for each, folks would likely still need to create custom reports to view data in any meaningful way. 
Keeping everything in one organization and ID also means we'll have one account to administer, instead of multiple, and we can help folks create reports easier. See @chalin's issue comment for more details on this organizational structure and links to google documentation on this: https://github.com/kubernetes/website/issues/37801#issuecomment-1309395460.

**Comment by chalin** [comment]
I'm going to close this issue since the main server and secondary servers have been migrated as of 2023/03/01. 🎉 

There are followup items being tracked via separate issues (see the opening comment).

Branch-release leads will need to ensure that GA4 is enabled in future releases. I leave further work in your capable hands. I'll be available if y'all need assistance. For the record, here were my attempts at this:

- #36325
- #36334

Finally, note that **as of 2024Q1 UA data will be deleted**. #sig-docs-maintainers should be aware of this so that they can plan, if necessary, a strategy for backing up the data.
